onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -124,79 +123,28 @@ void AddBinaryOperator(int32_t op_type,
   model_builder.AddOperation(op_type, input_indices, {output}, {output_operand_type}, {output_is_nhwc});
 }
 
-int GetType(const ONNX_NAMESPACE::ModelProto& model_proto,
-            const std::string& name) {
-  int invalid_type = ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;
-  for (const auto& input : model_proto.graph().input()) {
-    if (input.name() != name)
-      continue;
-
-    return input.type().tensor_type().elem_type();
-  }
-
-  for (const auto& value_info : model_proto.graph().value_info()) {
-    if (value_info.name() != name)
-      continue;
-
-    if (!value_info.has_type()) {
-      return invalid_type;
-    } else if (!value_info.type().has_tensor_type()) {
-      return invalid_type;
-    }
-
-    return value_info.type().tensor_type().elem_type();
+bool GetType(const onnxruntime::Node& node, int32_t& type) {
+  type = ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;
+  auto* type_proto = node.InputDefs()[0]->TypeAsProto();
+  if (!type_proto || !type_proto->tensor_type().has_elem_type()) {
+    LOGS_DEFAULT(VERBOSE) << "[" << node.OpType() << "] has no input type";
+    return false;
   }
 
-  return invalid_type;
+  type = type_proto->tensor_type().elem_type();
+  return true;
 }
 
-Shaper::Shape GetShape(const ONNX_NAMESPACE::ModelProto& model_proto,
-                       const std::string& name) {
-  Shaper::Shape empty_shape;
-  for (const auto& input : model_proto.graph().input()) {
-    if (input.name() != name)
-      continue;
-
-    Shaper::Shape shape;
-    for (const auto& dim : input.type().tensor_type().shape().dim())
-      shape.push_back(dim.dim_value());
-
-    return shape;
-  }
-
-  for (const auto& tensor : model_proto.graph().initializer()) {
-    if (tensor.name() != name)
-      continue;
-
-    Shaper::Shape shape;
-    for (auto dim : tensor.dims())
-      shape.push_back(SafeInt<uint32_t>(dim));
-
+Shaper::Shape GetShape(const ONNX_NAMESPACE::TensorShapeProto* shape_proto) {
+  Shaper::Shape shape;
+  if (!shape_proto)
     return shape;
-  }
-
-  for (const auto& value_info : model_proto.graph().value_info()) {
-    if (value_info.name() != name)
-      continue;
-
-    if (!value_info.has_type()) {
-      return empty_shape;
-    } else if (!value_info.type().has_tensor_type()) {
-      return empty_shape;
-    } else if (!value_info.type().tensor_type().has_shape()) {
-      return empty_shape;
-    } else if (value_info.type().tensor_type().shape().dim_size() == 0) {
-      return empty_shape;
-    }
 
```

Comment:
This doesn't disambiguate between an empty shape (scalar) and missing shape info.